src/common/stack_mmap_noXStream.c

```diff
@@ -0,0 +1,50 @@
+/**
+ * (C) Copyright 2016-2021 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * This file contains the engine/XStream specific part of the
+ * mmap()'ed ULT stack allocation feature.
+ */
+
+#ifdef ULT_MMAP_STACK
+#include <daos/common.h>
+#include <daos/stack_mmap.h>
+#include <errno.h>
+#include <string.h>
+
+/* XXX both thresholds may need to be dynamically determined based on the
+ * number of free stack pool (one per-XStream), and max_nb_mmap_stacks (see
+ * below).
+ */
+#define MAX_PERCENT_FREE_STACKS 20
+#define MAX_NUMBER_FREE_STACKS 2000
+
+/* callback to free stack upon ULT exit during stack_key deregister */
+void free_stack(void *arg)
+{
+	mmap_stack_desc_t *desc = (mmap_stack_desc_t *)arg;
+	struct stack_pool *sp;
+
+	sp = desc->sp;
+	free_stack_in_pool(desc, sp);
+}
+
+/* wrapper for ULT main function, mainly to register mmap()'ed stack
+ * descriptor as ABT_key to ensure stack pooling or munmap() upon ULT exit
+ */
+void mmap_stack_wrapper(void *arg)
+{
+	mmap_stack_desc_t *desc = (mmap_stack_desc_t *)arg;
+
+	ABT_key_set(stack_key, desc);
+
+	D_DEBUG(DB_MEM,
+		"New ULT with stack_desc %p running on CPU=%d\n",
+		desc, sched_getcpu());
+	desc->thread_func(desc->thread_arg);
+}
+#endif
```

Comment:
Why a separate C file? Should we move these two functions into the 'stack_mmap.c'?

Reply:
I think there are two ways to solve this:

1. Introduce a callback for getting the 'stack_pool' for stack free, if the callback is provided, call the callback to get 'stack_pool' for stack free, otherwise, use the 'stack_pool' in 'mmap_stack_desc_t'. DAOS engine provides the callback which calls dss_current_xstream() to get the 'stack_pool' for free. 
2. Pass two 'stack_pool's to mmap_stack_thread_create_common(), one for allocating stack, the other for freeing stack.

Reply:
hmmm, you are be right it should work. But I understand it is only to remove the need of the 2x stack_mmap_[no]XStream.c files and to fix the current watchdog mechanism compliance (that I think I can fix in an other way, see my related comment later), right ?! Since @johannlombardi agrees with you that the scenario of an ULT migrating between XStreams do (will ?) not exists in DAOS, I will give a try to one of your static solutions and get back to you after some required testing.